src/radical/pilot/agent/executing/base.py

```diff
@@ -117,7 +125,54 @@ def work(self, tasks):
     #
     def command_cb(self, topic, msg):
 
-        raise NotImplementedError('work is not implemented')
+        raise NotImplementedError('command_cb is not implemented')
+
+
+    # --------------------------------------------------------------------------
+    #
+    def cancel_task(self, task):
+
+        raise NotImplementedError('cancel_task is not implemented')
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _to_watcher(self):
+        '''
+        watch the set of tasks for which timeouts are defined.  If the timeout
+        passes and the tasks are still active, kill the task via
+        `self._cancel_task(task)`.  That has to be implemented by al executors.
+        '''
+
+        while True:
+
+            # check once per second at most
+            time.sleep(1)
+
+            now = time.time()
+            with self._to_lock:
+
+                # running tasks for next check
+                to_list = list()
+                for to, start, task in self._to_tasks:
+                    if now - start > to:
+                        self._prof.prof('task_timeout', uid=task['uid'])
+                        self.cancel_task(task)
+                    else:
+                        to_list.append([to, start, task])
+
+                self._to_tasks = to_list
```

Comment:
I would propose not to create a new list every time we go through the tasks (no need to make garbage collector's job harder :) )

either
```python
n_to_tasks = len(self._to_tasks)
while n_to_tasks:
    to, start, task = self._to_tasks.pop(0)
    if now - start > to:
        ...
    else:
        self._to_tasks.append([to, start, task])
    n_to_tasks -= 1
```
or just go through each element and remove cancelled tasks only
```python
idx = 0
for _ in range(len(self._to_tasks)):
    to, start, task = self._to_tasks[idx]
    if now - start > to:
        ...
        del self._to_tasks[idx]
    else:
        idx += 1
```

Reply:
I disagree with that - don't optimize before measuring, and even then, make sure it's worth the added complexity (we use the scheme above in many places).

Reply:
```py
#!/usr/bin/env python3

import time
import random

def test_1(n):
    items = list(range(n))
    n_items = len(items)
    while n_items:
        item = items.pop(0)
        if random.choice([True, False]):
            pass
        else:
            items.append(item)
        n_items -= 1


def test_2(n):
    items = list(range(n))
    idx = 0
    for _ in range(len(items)):
        item = items[idx]
        if random.choice([True, False]):
            del items[idx]
        else:
            idx += 1


def test_3(n):
    items = list(range(n))
    new_items = list()
    for item in items:
        if random.choice([True, False]):
            new_items.append(item)

    items = new_items


if __name__ == '__main__':

    for m in [1, 2, 4, 8, 16, 32, 64, 128, 256, 512]:

        n = 1024 * m

        start = time.time()
        test_1(n)

        print('%6dk : ' % m, end='')
        print(' %7.1f' % (time.time() - start), end='')

        start = time.time()
        test_2(n)
        print(' %7.1f' % (time.time() - start), end='')

        start = time.time()
        test_3(n)
        print(' %7.1f' % (time.time() - start), end='')

        print()
```

```sh
$ ./t.py 
     1k :      0.0     0.0     0.0
     2k :      0.0     0.0     0.0
     4k :      0.0     0.0     0.0
     8k :      0.0     0.0     0.0
    16k :      0.0     0.0     0.0
    32k :      0.2     0.1     0.0
    64k :      0.7     0.3     0.0
   128k :      3.0     1.0     0.1
   256k :     11.0     3.7     0.1
   512k :     45.0    15.7     0.3
```